In [8]:
# Imports section
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline

## Part 1. Loading the dataset

In [3]:
# Using pandas load the dataset (load remotely, not locally)
url = 'https://raw.githubusercontent.com/COMPSCI-383-Fall2024/assignment-4-equation-of-a-slime-<redacted>/refs/heads/main/science_data_large.csv?token=GHSAT0AAAAAACXWOIZHGSMPKEXP7PV4XWTEZYSR5ZA'
dataset = pd.read_csv(url)
# Output the first 15 rows of the data
print(dataset.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(dataset.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset.drop(columns=['Size nm^3'])
y = dataset['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)

## Part 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set
model = LinearRegression()
# Create a sample datapoint and predict the output of that sample with the trained model
model.fit(X_train, y_train)
sample_datapoint = [[400, 690]] 
predicted_output = model.predict(sample_datapoint)
print(predicted_output)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
r2 = model.score(X_test, y_test)
print(f"R² Score: {r2:.5f}")
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
beta0 = intercept
beta1, beta2 = coefficients
latex_equation = f"h(x) = {beta0:.5f} + {beta1:.5f} * Temperature + {beta2:.5f} * Mols KCL"
print(latex_equation)

[639354.4410871]
R² Score: 0.87616
h(x) = -400305.91333 + 863.58109 * Temperature + 1006.12742 * Mols KCL


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


The R² score, or coefficient of determination, measures the proportion of variance in the dependent variable that is predictable from the independent variables in a regression model. An R² score close to 1 indicates that a significant portion of the variance is explained by the model. 

h(x) = $\beta_0 + \beta_1 \cdot \text{Temperature} + \beta_2 \cdot \text{Mols KCL}$


## Part 4. Use Cross Validation

In [7]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cv_scores = cross_val_score(model, X, y, cv=10, scoring='r2')
mean_score = cv_scores.mean()
std_score = cv_scores.std()
# Report on their finding and their significance
print(f"Mean R² score from cross-validation: {mean_score:.5f}")
print(f"Standard deviation of R² score: {std_score:.5f}")

Mean R² score from cross-validation: 0.85525
Standard deviation of R² score: 0.03153


If the mean R² score is high (ours is 0.85525, which is close to 1), it suggests that the linear regression model is likely a good fit for the data, which supports the conclusions drawn from the earlier analyses.
The standard deviation gives insight into the model's robustness; ideally, you want to see a low standard deviation (ours is 0.03153, which is low). This means that no specific subsets of data are influencing the model's predictive capability significantly more than others.

## Part 5. Using Polynomial Regression

In [12]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
poly_model = LinearRegression()
poly_model.fit(X_poly, y)
r2_poly = poly_model.score(poly.transform(X_test), y_test)
print(f"R² Score for Polynomial Regression: {r2_poly:.5f}")
# Report on the metrics and output the resultant equation as you did in Part 3.

R² Score for Polynomial Regression: 1.00000


Polynomial regression of degree 2 models the relationship between the dependent variable and independent variables as a quadratic equation, allowing for both linear and quadratic terms, as well as interaction terms. The R² score for the polynomial regression indicates the proportion of variance in the dependent variable that is explained by the independent variables; a higher R² score (ours is 1) demonstrates a better fit of the model to the data.